src/wallet/wallet2.h

```diff
@@ -508,6 +514,7 @@ namespace tools
         FIELD(unlock_time)
         FIELD(use_rct)
         FIELD(rct_config)
+        FIELD(use_view_tags)
```

Comment:
Can this change in serialization make any things stored in files suddenly and unexpectedly invalid? And make them unrecoverable / unusable that way? Did you get an overwiew when working on this PR?

Reply:
You are right, this presents an issue worth discussing. Still looking into the full ramifications. Here's what I have so far:

AFAICT this serialization change renders exported txs incompatible across versions, and in particular: unsigned tx's. Even before the fork height, this would affect multisig, offline signing, and hw wallets relying on passing around unsigned tx's between wallet software and hw wallet (all of them?). For example, I can't export tx's from an older wallet and sign with a newer wallet or vise versa, multisig participants must all be using the same wallet version, a user must use a compatible hw wallet and wallet software etc. 

I'm not sure of a way around this issue. Unfortunately the pattern that seems like it could have been used to minimize the damage of a breaking issue like this with the other nearby structs isn't present on this struct (the `VERSION_FIELD`). Looks like that pattern was implemented later on after this struct was already set in stone. Probably makes sense to use this as an opportunity to add a `VERSION_FIELD` here too for future breaking changes on this struct.

I'm also not sure if this would break more things I'm not seeing (possibly the RPC wallet -- still looking into it).
_______

For the record, I could update to the new code using an existing old wallet file, no issue. After receiving tx's with view tags, I can't load the wallet file with the older software version. Figure that's expected.

Reply:
Update: exhaustively investigating a way to not cause a breaking change here in the least hacky way possible

Reply:
Splendid!

I am not sure yet whether I will just wait for your results, or play around with this a little myself as well. I learned a thing or two about that serialization stuff when analyzing the tag serialization question and may also try something just for learning more, maybe tomorrow.

Reply:
What I ask myself right now after a closer look: Why is that new field `use_view_tags` needed and essential in the first place? As far as I could see it will only gets used later when signing. What can go wrong if we drop this field here (again), and when the time to sign finally comes, maybe much later, we make an independent and **fresh** decision only then whether to use / generate a view tag or not for the final, fully constructed tx?

Thus I put up one of those "Change my mind" meme signs: I clain we don't need this. Change my mind :)

Reply:
When offline signing e.g., you don't know what height the chain is at and can't query for it because you're offline, so you don't know if the tx needs view tags or not. This is the design functionality the downstream consumers of this struct maintain. It's gotta be there at some level so they know unfortunately.

I'm rounding out testing of what I think is a solid implementation now, almost done :)

Reply:
Good argument, but in this case I am confused: If the software that signs offline does not know the chain height, how does it know what tx version to use for tx construction? I mean tx version on a much more fundamental level than just view tags yes/no, with ring size, range proof type to use, etc. Is this all contained somewhere in this construction data?

And if yes, wouldn't it be possible for you to "piggy-back" somewhere there, e.g. checking `RCTConfig.range_proof_type`?

But yeah, if you really find a good way to extend this in a compatible way and even sneak in a `VERSION_FIELD` without really breaking anything there would be no need to resort to little hacks like banking on the fact that the range proofs will change as well with this hardfork.

Reply:
[Here it is :)](https://github.com/monero-project/monero/pull/8061/commits/3f5e3486ae82a850712537bd47f73a91cc8e3b82)

I asked @moneromoo if they think there's a good way to avoid the break. moo suggested going with the breaking change as it originally was before this change, and adding a `VERSION_FIELD` to this struct for future changes. But, moo also came up this super clever idea to not cause a break, and I really like it.

I think this idea is the best maneuver to avoid a breaking change; I didn't think of anything close to as nice.

Its downside is that the serialized `use_rct` field is changing types from `bool` to `uint8_t`, and is now re-interpreted as `construction_flags` in the code (which confusingly, includes `use_rct` as one of the flags in addition to `use_view_tags` now). This naming inconsistency will naturally be a little confusing, and the `construction_flags` type is less intuitive than using booleans, but I'm not seeing issues with this. Wallets need to support view tags when the fork comes, so I figure they will need to update to use this new type in the correct way to be ready for the fork anyway. In the meantime, compatibility for users who don't update won't be broken right when v0.18.0.0 is released.

Simply breaking the struct would avoid the downsides with this `construction_flags` approach, and allows us to potentially incur some cost today for a better future. But, with such a large change on the horizon in Seraphis, perhaps we save our breaking change energy for that.

I'm fine with reverting + adding `VERSION_FIELD` if this case is weak. I don't want to hold up view tags much longer over this.
_________

> If the software that signs offline does not know the chain height, how does it know what tx version to use for tx construction? ... Is this all contained somewhere in this construction data?

Yep

https://github.com/monero-project/monero/blob/9aab19f349433687c7aaf2c1cbc5751e5912c0aa/src/cryptonote_core/cryptonote_tx_utils.cpp#L220

That `rct` is the `use_rct` boolean

> wouldn't it be possible for you to "piggy-back" somewhere there, e.g. checking RCTConfig.range_proof_type

Problem with this is we still support `use_view_tags` true and `use_rct` false, i.e. can still make tx's without the range proof type set (`sweep_unmixable` in the CLI will do this)

One solution I was exploring was relying on `rct_config`'s `VERSION_FIELD` and placing `use_view_tags` in there. I don't like it because `rct_config` is used across RingCT code, and it would be confusing to place anything to do with view tags inside there because it doesn't have anything to do with RingCT.